### PR TITLE
Fix Yul IR push with and without argument

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes https://github.com/ethereum/solidity/issues/7840

### CODE REVIEW

File: libsolidity/codegen/YulUtilFunctions.h [hunk 1]
```diff
@@ -126,9 +126,13 @@ class YulUtilFunctions
 	std::string storageArrayPopFunction(ArrayType const& _type);
 
 	/// @returns the name of a function that pushes an element to a storage array
-	/// signature: (array, value) -> newlength
+	/// signature: (array, value)
 	std::string storageArrayPushFunction(ArrayType const& _type);
 
+	/// @returns the name of a function that pushes the base type's zero element to a storage array
```
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.h]: `and returns a reference to the added element.`

File: test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul [hunk 2]
```diff
@@ -12,7 +12,7 @@
 // {
 //     {
 //         let y := mload(0x20)
-//         for { } iszero(iszero(and(y, 8))) { if y { revert(0, 0) } }
+//         for { } and(y, 8) { if y { revert(0, 0) } }
```
Comment (leonardoalt) [inline test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul:15]: How come?

**Review by leonardoalt — COMMENTED** [review]
Looks good, just some small comments

File: test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul (hunk 2 above)
Comment (ekpyron) [reply test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul:15]: Broken on ``develop_060`` in general... we might want to fix it in a separate PR... it's a separate commit here at least.

**Comment by ekpyron** [comment]
@leonardoalt Regarding the title: it returned the wrong stuff, if called without argument, we just also never tested what it returns :-). So this PR actually fixes both versions...

**Comment by leonardoalt** [comment]
Ah ok ;)

File: test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul (hunk 2 above)
Comment (leonardoalt) [reply test/libyul/yulOptimizerTests/fullSuite/clear_after_if_continue.yul:15]: ok

**Review by leonardoalt — APPROVED** [review]
Looks good if tests pass

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -659,15 +659,28 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
-				m_utils.zeroValueFunction(*arrayType.baseType()) + "()" :
-				expressionAsType(*arguments.front(), *arrayType.baseType())
-			) << ")\n";
+		if (arguments.empty())
+		{
+			auto slotName = m_context.newYulVariable();
+			auto offsetName = m_context.newYulVariable();
+			m_code << "let " << slotName << ", " << offsetName << " := " <<
+				m_utils.storageArrayPushZeroFunction(arrayType) <<
+				"(" << m_context.variable(_functionCall.expression()) << ")\n";
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:668]: Type conversion?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 4]
```diff
@@ -659,15 +659,28 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+				slotName,
+				offsetName,
+				*arrayType.baseType()
+			));
+		}
+		else
+			m_code <<
+				m_utils.storageArrayPushFunction(arrayType) <<
+				"(" <<
+				m_context.variable(_functionCall.expression()) <<
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:680]: Type conversion?

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:680]: Ah no, it's the function call!

**Review by chriseth — APPROVED** [review]

**Pull request merged**